src/NServiceBus.Transport.Msmq/DelayedDelivery/DueDelayedMessagePoller.cs

```diff
@@ -0,0 +1,341 @@
+
+        async Task WaitForSignal(CancellationToken cancellationToken)
+        {
+            try
+            {
+                await signalQueue.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+            {
+                //Shutting down
+            }
+        }
+
+        Task WaitIfNeeded(DateTimeOffset nextPoll, CancellationToken cancellationToken)
+        {
+            var waitTime = nextPoll - DateTimeOffset.UtcNow;
+            if (waitTime > TimeSpan.Zero)
+            {
+                return Task.Delay(waitTime, cancellationToken);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        async Task Poll(TaskCompletionSource<DateTimeOffset?> result, CancellationToken cancellationToken)
+        {
+            DelayedMessage timeout = null;
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+            try
+            {
+                using (var tx = new TransactionScope(TransactionScopeOption.Required, transactionOptions, TransactionScopeAsyncFlowOption.Enabled))
+                {
+                    timeout = await delayedMessageStore.FetchNextDueTimeout(now, cancellationToken).ConfigureAwait(false);
+                    fetchCircuitBreaker.Success();
+
+                    if (timeout != null)
+                    {
+                        result.SetResult(null);
+                        await HandleDueDelayedMessage(timeout, cancellationToken).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        using (new TransactionScope(TransactionScopeOption.RequiresNew, transactionOptions, TransactionScopeAsyncFlowOption.Enabled))
+                        {
+                            var nextDueTimeout = await delayedMessageStore.Next(cancellationToken).ConfigureAwait(false);
+                            if (nextDueTimeout.HasValue)
+                            {
+                                result.SetResult(nextDueTimeout);
+                            }
+                            else
+                            {
+                                //If no timeouts, wait a while
+                                result.SetResult(now.Add(MaxSleepDuration));
+                            }
+                        }
+                    }
+
+                    tx.Complete();
+                }
+            }
+            catch (QueueNotFoundException exception)
+            {
+                if (timeout != null)
+                {
+                    await TrySendDelayedMessageToErrorQueue(timeout, exception, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+            {
+                //Shutting down
+                throw;
+            }
+            catch (Exception exception)
```

Comment:
If we are shutting down shouldn't this then be a `return`?

Otherwise, maybe an improved comment:
```suggestion
            // Prevent the `catch(Exception)` to do any tasks in case of shutdown.
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                throw;
            }
            catch (Exception exception)
```

Reply:
I based this code on https://docs.particular.net/nservicebus/cancellation-and-catching-exceptions I added some code to `AwaitHandleTasks` to handle this `throw` but you are right it could be replaced with a `return` and `AwaitHandleTasks` would not have to handle it. 

I am fine either way, I am not an expert in cancellation.

Reply:
@SzymonPobiega it just reads strange as that catch block only does a throw which would be the default when not having catch. Hence adding a comment to ensure the intent it easily understood.